### ISSUE TITLE
Run GoodJob on puma boot

### ISSUE THREAD

**Issue by sj26** [bensheldon/good_job#91]
I'm running an app on Heroku and want to use a single dyno, and good_job is fitting the bill nicely. I want to run good_job in async mode on puma boot. But I don't want to run a new good_job async if I start a rails console, so I can't put it in my `config/environments/production.rb`. I also noticed that just setting `queue_adapter` to `:good_job` and then setting `GOOD_JOB_EXECUTION_MODE=async` wasn't enough to reliably start good job on puma boot until an activejob was touched (and autoloading configured [and started] the queue adapter). So I ended up using:

```ruby
# config/application.rb

module MyGreatApp
  class Application < Rails::Application
    # ...

    config.active_job.queue_adapter = :good_job
  end
end
```

```ruby
# config/puma.rb

# Make sure good_job runs in the puma process as soon as puma is started, and
# always -- even in production
# 
# Note if `workers` is set then this will need to be wrapped in `on_worker_boot`
Rails.application.config.active_job.queue_adapter = GoodJob::Adapter.new(execution_mode: :async)
```

This means the autoloading behaviour is all good, in development it defaults to running inline but also has an async processor running in puma which is processing immediately, and always starts an async processor in the puma process in production while using external from rails console for enqueuing jobs.

This seemed like a useful little journey worth sharing. :-)

**Comment by bensheldon** [comment]
Thank you for documenting this 🙏  I think you're right that configuring the adapter inside of `config/puma.rb` works best right now.

I will dig into #89 too. I'm thinking that it may be safest to defer starting the background processing (e.g. the Notifier and Scheduler) until after the application has fully loaded via an initializer.

**Comment by sj26** [comment]
Thanks! Yeah, I'm wondering if good_job needs a railtie. Perhaps it can use that to hook into the lifecycle of the rails server command, or something.

**Comment by sj26** [comment]
Note that [auto_terminate](https://github.com/bensheldon/good_job/blob/main/lib/good_job/notifier.rb#L12) doesn't seem to have any effect in concurrent-ruby, so to get puma to reliably exit when running good_job async I had to change this to:

```
good_job = GoodJob::Adapter.new(execution_mode: :async)
Rails.application.config.active_job.queue_adapter = good_job
at_exit { good_job.shutdown }
```

**Comment by bensheldon** [comment]
The startup/initialization issues have been addressed in #199 and #205 

This using `if Rails.const_defined?("Server") ...` is a pattern that seems reliable at least on Puma: https://github.com/bensheldon/good_job/blob/b1cd2226a167a36a9194ed8e81dbfc16074e4117/spec/test_app/config/environments/development.rb#L63-L70

**Comment by bensheldon** [comment]
Also, this is current Readme documentation for Puma:

https://github.com/bensheldon/good_job/blob/b1cd2226a167a36a9194ed8e81dbfc16074e4117/README.md#L508-L531
